### PR TITLE
Add Tqdm Progress Bar to segment_volume

### LINKED ISSUE #1114: Add progress bar to segment_volume
### Motivation for the feature
With microscopy images, we and our users are dealing with fairly large images that can take a lot of time to segment (with the `segment_volume` API) and the progress ETA in the terminal is not clear.

At first there is the number of patches, slices or sub-volumes that are loaded:
```
2022-03-25 15:01:18.974 | INFO     | ivadomed.inference:segment_volume:451 - Loaded 20 axial patches of shape [256, 256].
```
And then a repetition of the following lines for each batch (according to batch size) of patches, slices or sub-volumes:
```
2022-03-25 15:01:18.983 | INFO     | ivadomed.utils:define_device:130 - Cuda is not available.
2022-03-25 15:01:18.983 | INFO     | ivadomed.utils:define_device:131 - Working on cpu.
2022-03-25 15:01:18.984 | DEBUG    | ivadomed.inference:get_preds:88 - Likely ONNX model detected at: <path_of_the_model>
2022-03-25 15:01:18.985 | DEBUG    | ivadomed.inference:get_preds:89 - Conduct ONNX model inference...
2022-03-25 15:01:21.127 | DEBUG    | ivadomed.inference:get_preds:92 - Sending predictions to CPU
```

### Description of the feature
We should add a display of the progress of the segmentation batches with `tqdm` as it is done with the `--test` command in [testing.py](https://github.com/ivadomed/ivadomed/blob/1a3d193922938dc35e9eb77d6af7fd8e6d1bacfa/ivadomed/testing.py#L123).

### PR DESCRIPTION
**Pull request by cakester**
#### GitHub

- [ ] I've given this PR a concise, self-descriptive, and meaningful title
- [ ] I've linked relevant issues in the PR body
- [ ] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [ ] I've assigned a reviewer

#### PR contents

- [ ] I've consulted [ivadomed's internal developer documentation](https://github.com/ivadomed/ivadomed/wiki) to ensure my contribution is in line with any relevant design decisions
- [ ] I've added [relevant tests](https://github.com/ivadomed/ivadomed/wiki/tests) for my contribution
- [ ] I've updated the [relevant documentation](https://github.com/ivadomed/ivadomed/wiki/documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description

Add tqdm progress bar to segment_volume

## Linked issues

resolves #1114

### CODE REVIEW

**Review by mariehbourget — COMMENTED** [review]
Thanks @cakester for taking care of this.
I think I would go with "Inference" instead of "Segment_volume" for the description.

I tested it in `ivadomed` with the `--segment` command and it works well.
However, when I test it with the `segment_volume` API through ADS, the progress bar doesn't show correctly. I'm not sure why, maybe because in ADS it is nested in another tqdm loop?
It looks like this:
image

So I tried adding the argument `position=0` in `ivadomed` and it works well.
image


I don't know if that is the "right" fix. Let me know what you think.

Also, as `segment_volume` is used by SCT as well, we should check with @joshuacwnewton to make sure the change is OK on his side.

**Comment by cakester** [comment]
what command did you run to test segment_volume through ADS?

**Comment by mariehbourget** [comment]
> what command did you run to test segment_volume through ADS?

Sorry, I should have include that in my comment!

I have ADS installed in its own conda environment (it includes the latest version of ivadomed as a dependency): https://axondeepseg.readthedocs.io/en/latest/documentation.html#installation
Then you can run ADS with: https://axondeepseg.readthedocs.io/en/latest/documentation.html#segment-a-single-image
On this data: https://axondeepseg.readthedocs.io/en/latest/documentation.html#example-dataset

To test the fix, I manually added the `position=0` argument in `inference.py` in `miniconda\envs\ads_venv_v4\lib\python3.8\site-packages\ivadomed`.

File: ivadomed/inference.py [hunk 1]
```diff
@@ -468,7 +469,7 @@ def segment_volume(folder_model: str, fname_images: list, gpu_id: int = 0, optio
     # Loop across batches
     preds_list, slice_idx_list = [], []
     last_sample_bool, weight_matrix, volume, image = False, None, None, None
-    for i_batch, batch in enumerate(data_loader):
+    for i_batch, batch in enumerate(tqdm(data_loader, desc="Segment_volume")):
```
[the remark this answers is not in the document]
Comment (joshuacwnewton) [reply ivadomed/inference.py:472]: > Also, as segment_volume is used by SCT as well, we should check with @joshuacwnewton to make sure the change is OK on his side.

As far as I can tell, SCT mostly uses `segment_volume` for single-volume segmentation, so the tqdm bar only ever shows as 1/1. 

Still, here's the before and after for the `position=0` change, if you're curious.

Before:

```console
joshua@tadpole:~/repos/spinalcordtoolbox/data/sct_example_data/t2s$ sct_deepseg -i t2s.nii.gz -task seg_sc_t2star -o t2s_seg_deepseg_tqdm.nii.gz -thr 0.9


--
Spinal Cord Toolbox (git-master-6b8613a10caa95d35e015ab35ca6e368ef067894)

sct_deepseg -i t2s.nii.gz -task seg_sc_t2star -o t2s_seg_deepseg_tqdm.nii.gz -thr 0.9
--

2022-04-08 16:37:13.367 | WARNING  | ivadomed.inference:segment_volume:404 - fname_roi has not been specified, then the entire volume is processed.
2022-04-08 16:37:13.866 | INFO     | ivadomed.inference:segment_volume:454 - Loaded 32 axial slices.
Segment_volume:   0%|                                     | 0/1 [00:00<?, ?it/s]2022-04-08 16:37:13.876 | INFO     | ivadomed.utils:define_device:131 - Cuda is not available.
2022-04-08 16:37:13.876 | INFO     | ivadomed.utils:define_device:132 - Working on cpu.
2022-04-08 16:37:13.876 | DEBUG    | ivadomed.inference:get_preds:72 - PyTorch model detected at: /home/joshua/repos/spinalcordtoolbox/data/deepseg_models/t2star_sc/t2star_sc.pt
2022-04-08 16:37:13.876 | DEBUG    | ivadomed.inference:get_preds:73 - Loading model from: /home/joshua/repos/spinalcordtoolbox/data/deepseg_models/t2star_sc/t2star_sc.pt
/home/joshua/repos/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/torch/serialization.py:657: SourceChangeWarning: source code of class 'ivadomed.models.Unet' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
/home/joshua/repos/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/torch/serialization.py:657: SourceChangeWarning: source code of class 'ivadomed.models.Encoder' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
/home/joshua/repos/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/torch/serialization.py:657: SourceChangeWarning: source code of class 'torch.nn.modules.container.ModuleList' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
/home/joshua/repos/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/torch/serialization.py:657: SourceChangeWarning: source code of class 'ivadomed.models.DownConv' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
/home/joshua/repos/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/torch/serialization.py:657: SourceChangeWarning: source code of class 'torch.nn.modules.conv.Conv2d' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
/home/joshua/repos/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/torch/serialization.py:657: SourceChangeWarning: source code of class 'ivadomed.models.Decoder' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
/home/joshua/repos/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/torch/serialization.py:657: SourceChangeWarning: source code of class 'ivadomed.models.UpConv' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
2022-04-08 16:37:13.900 | DEBUG    | ivadomed.inference:get_preds:76 - Evaluating model: /home/joshua/repos/spinalcordtoolbox/data/deepseg_models/t2star_sc/t2star_sc.pt
2022-04-08 16:37:20.425 | DEBUG    | ivadomed.inference:get_preds:93 - Sending predictions to CPU
Segment_volume: 100%|█████████████████████████████| 1/1 [00:06<00:00,  6.81s/it]
```

After:

```
joshua@tadpole:~/repos/spinalcordtoolbox/data/sct_example_data/t2s$ sct_deepseg -i t2s.nii.gz -task seg_sc_t2star -o t2s_seg_deepseg_tqdm.nii.gz -thr 0.9

--
Spinal Cord Toolbox (git-master-6b8613a10caa95d35e015ab35ca6e368ef067894)

sct_deepseg -i t2s.nii.gz -task seg_sc_t2star -o t2s_seg_deepseg_tqdm.nii.gz -thr 0.9
--

2022-04-08 16:45:55.902 | WARNING  | ivadomed.inference:segment_volume:404 - fname_roi has not been specified, then the entire volume is processed.
2022-04-08 16:45:56.390 | INFO     | ivadomed.inference:segment_volume:454 - Loaded 32 axial slices.
Inference::   0%|                                                                  | 0/1 [00:00<?, ?it/s]2022-04-08 16:45:56.400 | INFO     | ivadomed.utils:define_device:131 - Cuda is not available.
2022-04-08 16:45:56.400 | INFO     | ivadomed.utils:define_device:132 - Working on cpu.
2022-04-08 16:45:56.400 | DEBUG    | ivadomed.inference:get_preds:72 - PyTorch model detected at: /home/joshua/repos/spinalcordtoolbox/data/deepseg_models/t2star_sc/t2star_sc.pt
2022-04-08 16:45:56.400 | DEBUG    | ivadomed.inference:get_preds:73 - Loading model from: /home/joshua/repos/spinalcordtoolbox/data/deepseg_models/t2star_sc/t2star_sc.pt
/home/joshua/repos/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/torch/serialization.py:657: SourceChangeWarning: source code of class 'ivadomed.models.Unet' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
/home/joshua/repos/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/torch/serialization.py:657: SourceChangeWarning: source code of class 'ivadomed.models.Encoder' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
/home/joshua/repos/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/torch/serialization.py:657: SourceChangeWarning: source code of class 'torch.nn.modules.container.ModuleList' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
/home/joshua/repos/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/torch/serialization.py:657: SourceChangeWarning: source code of class 'ivadomed.models.DownConv' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
/home/joshua/repos/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/torch/serialization.py:657: SourceChangeWarning: source code of class 'torch.nn.modules.conv.Conv2d' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
/home/joshua/repos/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/torch/serialization.py:657: SourceChangeWarning: source code of class 'ivadomed.models.Decoder' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
/home/joshua/repos/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/torch/serialization.py:657: SourceChangeWarning: source code of class 'ivadomed.models.UpConv' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
2022-04-08 16:45:56.423 | DEBUG    | ivadomed.inference:get_preds:76 - Evaluating model: /home/joshua/repos/spinalcordtoolbox/data/deepseg_models/t2star_sc/t2star_sc.pt
2022-04-08 16:46:02.701 | DEBUG    | ivadomed.inference:get_preds:93 - Sending predictions to CPU
Inference:: 100%|██████████████████████████████████████████████████████████| 1/1 [00:06<00:00,  6.56s/it]
```

Not much difference? 

(Aside: Looks like the extra `:` isn't needed for `Inference:`, as it's doubled-up in the output.

[the remark this answers is not in the document]
Comment (mariehbourget) [reply ivadomed/inference.py]: Thanks @joshuacwnewton for testing!

> As far as I can tell, SCT mostly uses `segment_volume` for single-volume segmentation, so the tqdm bar only ever shows as 1/1. (Unless there are other models that I can try that would handle things with some sort of patch-based inference? Maybe I'm testing this incorrectly. 🤔)

The denominator of tqdm is the number of "batches" that are loaded. If you segment 1 volume that has a number of 2D slices <= batchsize, you will always have 1. In microscopy, we have 2D patches in addition to 2D slices and small batches of size 4, so we almost have always multiple batches. To test, you can reduce the batchsize in the config file of the model you are using.

> Aside 2: Have those `SourceChangeWarnings` always been there?

I had never seen this warning, but I'm working with `onnx` models (CPU) most of the time. I tried with the `pt` model and got the same warning as you in python 3.8. I have no idea since when they have been there 🤔

**Comment by jcohenadad** [comment]
> As far as I can tell, SCT mostly uses segment_volume for single-volume segmentation, so the tqdm bar only ever shows as 1/1. (Unless there are other models that I can try that would handle things with some sort of patch-based inference? Maybe I'm testing this incorrectly. 🤔)

it would be great to have tqdm show progression across slices (in case of 2d kernels), as currently done in `sct_segment_gm`. And for 3d kernels, is there a way we can iterate across something?

**Comment by cakester** [comment]
> > what command did you run to test segment_volume through ADS?
> 
> Sorry, I should have include that in my comment!
> 
> I have ADS installed in its own conda environment (it includes the latest version of ivadomed as a dependency): https://axondeepseg.readthedocs.io/en/latest/documentation.html#installation Then you can run ADS with: https://axondeepseg.readthedocs.io/en/latest/documentation.html#segment-a-single-image On this data: https://axondeepseg.readthedocs.io/en/latest/documentation.html#example-dataset
> 
> To test the fix, I manually added the `position=0` argument in `inference.py` in `miniconda\envs\ads_venv_v4\lib\python3.8\site-packages\ivadomed`.

image
my tqdm bar shows up fine, maybe the color it shows is the same as your background that's why you cannot see
